internal/jimm/bootstrap/bootstrap_test.go

```diff
@@ -0,0 +1,137 @@
+	"testing"
+	"time"
+
+	qt "github.com/frankban/quicktest"
+	"github.com/frankban/quicktest/qtsuite"
+	"github.com/google/uuid"
+
+	"github.com/canonical/jimm/v3/internal/db"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/jimm/bootstrap"
+	"github.com/canonical/jimm/v3/internal/jobtracker"
+	"github.com/canonical/jimm/v3/internal/openfga"
+	"github.com/canonical/jimm/v3/internal/testutils/jimmtest"
+	"github.com/canonical/jimm/v3/pkg/api/params"
+)
+
+type bootstrapManagerSuite struct {
+	manager    *bootstrap.BootstrapManager
+	jobTracker *jobtracker.Tracker
+	adminUser  *openfga.User
+	db         *db.Database
+	ofgaClient *openfga.OFGAClient
+}
+
+func (s *bootstrapManagerSuite) Init(c *qt.C) {
+	db := &db.Database{
+		DB: jimmtest.PostgresDB(c, time.Now),
+	}
+	err := db.Migrate(context.Background())
+	c.Assert(err, qt.IsNil)
+
+	s.db = db
+
+	ofgaClient, _, _, err := jimmtest.SetupTestOFGAClient(c.Name())
+	c.Assert(err, qt.IsNil)
+
+	s.ofgaClient = ofgaClient
+
+	jobtracker, err := jobtracker.NewJobTracker(db, 1*time.Minute)
+	s.jobTracker = jobtracker
+	c.Assert(err, qt.IsNil)
+	s.manager, err = bootstrap.NewBootstrapManager(db, ofgaClient, jobtracker)
+	c.Assert(err, qt.IsNil)
+}
+
+func (s *bootstrapManagerSuite) TestGetBootstrapStatusAndLogs(c *qt.C) {
+	ctx := c.Context()
+	read := make(chan struct{})
+	defer close(read)
+	write := make(chan struct{})
+	defer close(write)
+	numLogs := 101
+	batchSize := 10
+	jobId, err := s.jobTracker.Run(ctx,
+		"bootstrap-job",
+		func(ctx context.Context) error {
+			jobId, ok := jobtracker.JobIdFromContext(ctx)
+			c.Check(ok, qt.IsTrue)
+
+			for i := range numLogs {
+				if i%batchSize == 0 && i > 0 {
+					write <- struct{}{} // Signal that a batch of logs has been written.
+					<-read              // Wait for the read before writing the next batch.
+				}
+				err := s.db.AddBootstrapLog(ctx, jobId, "bootstrap logs "+fmt.Sprint(rune(i)))
+				c.Check(err, qt.IsNil)
+			}
+			write <- struct{}{}
+			<-read
+			return nil
+		},
+		1*time.Minute,
+	)
+	c.Assert(err, qt.IsNil)
+	watermark := 0
+	for i := 0; i < numLogs/batchSize+1; i++ {
+		<-write // Wait for the batch of logs to be written.
+		response, err := s.manager.GetBootstrapStatusAndLogs(ctx, s.adminUser, jobId, watermark)
+		c.Assert(err, qt.IsNil)
+		logs := []string{}
+		for j := 0; j < int(math.Min(float64(batchSize), float64(numLogs-i*batchSize))); j++ {
+			logs = append(logs, "bootstrap logs "+fmt.Sprint(rune(i*batchSize+j)))
+		}
+		c.Check(response.Logs, qt.DeepEquals, logs)
+		c.Assert(response.Status, qt.Equals, string(dbmodel.StatusRunning))
+		watermark = response.Watermark
+		read <- struct{}{} // Signal it has been read.
+	}
+
+	// check last batch is empty.
+	response, err := s.manager.GetBootstrapStatusAndLogs(ctx, s.adminUser, jobId, watermark)
+	c.Assert(response.Status == string(dbmodel.StatusSuccessful) || response.Status == string(dbmodel.StatusRunning), qt.IsTrue)
```

Comment:
Shouldn't we only wait for it to be successful? Can pinch my polling code from the job tracker test